test-suite/test-sources.sh

```diff
@@ -0,0 +1,262 @@
+        return 0;
+    fi
+
+    # TODO: Remove this year-2020 workaround? Fresh master PRs should have Bug 5021 fix.
+    if test -n "${PULL_REQUEST_NUMBER}"
+    then
+        # Detect stale PR branches that were forked before the first codespell
+        # application and, hence, cannot be spellchecked without stepping on
+        # misspellings in base code. Any PR-specific misspellings of stale
+        # branches should be caught later, when testing the staged commit.
+        required_title='Bug 5021: Spelling errors fixed by running scripts/spell-check.sh'
+        if ! git log --grep "$required_title" "HEAD^1" | grep -q ^commit
+        then
+            echo "The base code of this PR does not support automated spelling checks."
+            return 0;
+        fi
+    fi
+
+    # To avoid flagging out-of-scope misspellings, only check modified files.
+    # This also speeds up tests and helps avoid slow-codespell timeouts.
+    local changed_files="`git diff --name-only ${STARTING_POINT}`"
+    if test -z "$changed_files"
+    then
+        echo "ERROR: Unable to determine which files have changed."
+        return 1; # be conservative until we learn why that is a bad idea
+    fi
+    echo "changed files: $changed_files"
+
+    run ./scripts/spell-check.sh $changed_files
+
+    if run git diff --word-diff --exit-code
+    then
+        return 0;
+    fi
+
+    echo_error "Spelling mistakes detected."
+    echo "The log above ends with a word diff showing suggested fixes."
+    echo "Please adjust scripts/codespell-whitelist.txt or fix spelling."
+
+    return 1;
+}
+
+# check for certain problems fixed by a ./scripts/source-maintenance.sh
+check_source_maintenance() {
+    if ! git diff --quiet
+    then
+        CHECK_OUTCOME_PHRASE="Skipped due to a dirty working directory"
+        return 0;
+    fi
+
+    local checker=./scripts/source-maintenance.sh
+
+    local copyrightOption='--check-and-update-copyright'
+    if ! grep -q -e $copyrightOption $checker
+    then
+        echo_warning "Skipping $checker checks because $checker does not support $copyrightOption"
+        CHECK_OUTCOME_PHRASE="Skipped due to missing $copyrightOption support"
+        return 0;
+    fi
+
+    # The OS may not provide the right version of the astyle package, but
+    # source formatting (by developers) is not yet enforced, so we run with
+    # whatever astyle is provided, abusing the fact that $checker skips
+    # formatting iff it can execute a binary called astyle but does not like
+    # astyle's version.
+
+    # Avoid distracting $checker warnings; TODO: Fix $checker instead.
+    touch boilerplate_fix.sed
+
+    if ! run $checker $copyrightOption no
+    then
+        echo_error "Running $checker modified sources"
+        CHECK_OUTCOME_PHRASE="Ignored $checker failure" # maybe overwritten below
+        # TODO: Require source-maintenance.sh application instead of ignoring this error.
```

Comment:
This TODO and its sister TODO at the very end of this function deserve a dedicated policy-changing PR (which can also be used for discussing whether we are ready for requiring source-maintenance.sh application from contributors).